rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/core/ReconnectableRSocket.kt

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2015-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.rsocket.kotlin.core
+
+import io.ktor.utils.io.core.*
+import io.rsocket.kotlin.*
+import io.rsocket.kotlin.payload.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.*
+
+internal typealias ReconnectPredicate = suspend (cause: Throwable, attempt: Long) -> Boolean
+
+@Suppress("FunctionName")
+@OptIn(ExperimentalCoroutinesApi::class, FlowPreview::class)
+internal suspend fun ReconnectableRSocket(
+    connect: suspend () -> RSocket,
+    predicate: ReconnectPredicate,
+): RSocket {
+    val state = MutableStateFlow<ReconnectState>(ReconnectState.Connecting)
+
+    val job =
+        connect.asFlow()
+            .map<RSocket, ReconnectState> { ReconnectState.Connected(it) } //if connection established - state = connected
+            .onStart { emit(ReconnectState.Connecting) } //init - state = connecting
+            .retryWhen { cause, attempts -> predicate(cause, attempts) } //reconnection logic
+            .catch { emit(ReconnectState.Failed(it)) } //reconnection failed - state = failed
+            .mapNotNull {
+                state.value = it //set state //TODO replace with Flow.stateIn when coroutines 1.4.0 will be released
+
+                when (it) {
+                    is ReconnectState.Connected -> it.rSocket.join()  //await for connection completion
+                    is ReconnectState.Failed    -> throw it.error //reconnect failed, cancel job
+                    ReconnectState.Connecting   -> null //ignore, should never happen
```

Comment:
Suggest at minimum make some debugging hook here.  

```
ReconnectState.Connecting   -> impossible()
```

You could make impossible throw an assertion error, defined to return nothing, or at least a soft error like null with a log?

Reply:
need to adapt comment, it was there in first try of implementing it, and now it happens here, will do